File: tests/Blokemon.Game.Tests/TriggerTimingTests.cs
using Blokemon.Game;
using Shouldly;

namespace Blokemon.Game.Tests;

public sealed class TriggerTimingTests
{
    [Test]
    public async Task RuleBoxKnockout_TakesBothBarChitsAndWins()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-026",
            "BLK-151",
            ["VIM-BEER", "VIM-BEER", "VIM-SOBER"],
            97
        );
        var firstPrize = MatchScenario.Card(
            "prize-1",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        var secondPrize = MatchScenario.Card(
            "prize-2",
            "VIM-SOBER",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            1
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state.Cards.Append(firstPrize).Append(secondPrize).OrderBy(static card => card.Id)
            ),
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(player =>
                    player.Id == MatchScenario.FirstPlayer
                        ? player with
                        {
                            BarChitsRemaining = 2,
                        }
                        : player
                )
            ),
        };

        var applied = (CommandOutcome.Applied)
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-026-B01"));

        applied.State.Card(firstPrize.Id).Zone.ShouldBe(CardZone.Mitt);
        applied.State.Card(secondPrize.Id).Zone.ShouldBe(CardZone.Mitt);
        applied.State.Player(MatchScenario.FirstPlayer).BarChitsRemaining.ShouldBe(0);
        applied.State.Winner.ShouldBe(MatchScenario.FirstPlayer);
    }

    [Test]
    public async Task WouldBeKnockedOutTrigger_ResolvesBeforeKnockout()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-076",
            "BLK-068",
            ["VIM-LAIRY", "VIM-SOBER", "VIM-SOBER"],
            0
        );

        var applied = (CommandOutcome.Applied)
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-076-B02"));

        var defender = applied.State.Card(new CardInstanceId("defender"));
        defender.Zone.ShouldBe(CardZone.Oche);
        defender.Damage.ShouldBe(170);
    }

    [Test]
    public async Task DamagedActiveTrigger_PlacesCountersOnTheAttackerBeforeKnockouts()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState("BLK-076", "BLK-107", ["VIM-LAIRY"], 107);

        var applied = (CommandOutcome.Applied)
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-076-B01"));

        applied.State.Card(new CardInstanceId("attacker")).Damage.ShouldBe(30);
    }

    [Test]
    public async Task TakenFaceDownTrigger_WaitsForThePrizeOwnerAndCpuCompletesIt()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-003",
            "BLK-001",
            ["VIM-BLAZED", "VIM-BLAZED", "VIM-SOBER"],
            0
        );
        var triggeredPrize = MatchScenario.Card(
            "triggered-prize",
            "BLK-113",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        var extraPrize = MatchScenario.Card(
            "extra-prize",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            1
        );
        var defenderBench = MatchScenario.Card(
            "defender-bench",
            "BLK-004",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state
                    .Cards.Append(triggeredPrize)
                    .Append(extraPrize)
                    .Append(defenderBench)
                    .OrderBy(card => card.Id)
            ),
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(player =>
                    player.Id == MatchScenario.FirstPlayer
                        ? player with
                        {
                            BarChitsRemaining = 2,
                        }
                        : player
                )
            ),
        };

        var attacked = (CommandOutcome.Applied)
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-003-B01"));

        attacked.State.Phase.ShouldBe(MatchPhase.AwaitingTriggerChoice);
        attacked.State.PendingBarChits.Count.ShouldBe(1);

        var cpu = new DeterministicCpu();
        var decision = (CpuDecision.Selected)
            cpu.Choose(engine, attacked.State, MatchScenario.FirstPlayer);
        var resolved = (CommandOutcome.Applied)
            engine.Apply(attacked.State, decision.Action.Command);

        decision.Action.Kind.ShouldBe(LegalActionKind.ResolveBarChitTrigger);
        resolved.State.Card(triggeredPrize.Id).Zone.ShouldBe(CardZone.Booth);
        resolved.State.Card(extraPrize.Id).Zone.ShouldBe(CardZone.Mitt);
        resolved.State.Winner.ShouldBe(MatchScenario.FirstPlayer);
    }

    [Test]
    public async Task KnockoutEnergyMove_WaitsForItsOwnerBeforeKnockoutCompletes()
    {
        var engine = MatchScenario.Engine();
        var state = MatchScenario.BattleState(
            "BLK-003",
            "BLK-001",
            ["VIM-BLAZED", "VIM-BLAZED", "VIM-SOBER"],
            103
        );
        var triggerSource = MatchScenario.Card(
            "trigger-source",
            "BLK-026",
            MatchScenario.SecondPlayer,
            CardZone.Booth,
            -1
        );
        var movableVim = MatchScenario.Card(
            "movable-vim",
            "VIM-BEER",
            MatchScenario.SecondPlayer,
            CardZone.Attached,
            -1,
            attachedTo: new CardInstanceId("defender")
        );
        var prize = MatchScenario.Card(
            "prize",
            "VIM-LAIRY",
            MatchScenario.FirstPlayer,
            CardZone.BarChit,
            0
        );
        state = state with
        {
            Cards = FrozenList<CardState>.Create(
                state
                    .Cards.Select(card =>
                        card.Id.Value == "defender"
                            ? card with
                            {
                                Attachments = FrozenList<CardInstanceId>.Create(movableVim.Id),
                            }
                            : card
                    )
                    .Append(triggerSource)
                    .Append(movableVim)
                    .Append(prize)
                    .OrderBy(card => card.Id)
            ),
            Players = FrozenList<PlayerState>.Create(
                state.Players.Select(player =>
                    player.Id == MatchScenario.FirstPlayer
                        ? player with
                        {
                            BarChitsRemaining = 1,
                        }
                        : player
                )
            ),
        };

        var attacked = (CommandOutcome.Applied)
            engine.Apply(state, MatchScenario.AttackCommand(state, "BLK-003-B01"));

        attacked.State.Phase.ShouldBe(MatchPhase.AwaitingTriggerChoice);
        attacked.State.PendingKnockout.ShouldNotBeNull();
        attacked.State.Card(movableVim.Id).AttachedTo.ShouldBe(new CardInstanceId("defender"));
        attacked.State.Card(new CardInstanceId("defender")).Zone.ShouldBe(CardZone.Oche);

        var cpu = new DeterministicCpu();
        var decision = (CpuDecision.Selected)
            cpu.Choose(engine, attacked.State, MatchScenario.SecondPlayer);
        var resolved = (CommandOutcome.Applied)
            engine.Apply(attacked.State, decision.Action.Command);

        decision.Action.Kind.ShouldBe(LegalActionKind.ResolveKnockoutTrigger);
        resolved.State.Card(movableVim.Id).AttachedTo.ShouldBe(triggerSource.Id);
        resolved.State.Card(new CardInstanceId("defender")).Zone.ShouldBe(CardZone.EmptiesTray);
        resolved.State.Card(prize.Id).Zone.ShouldBe(CardZone.Mitt);
    }
}

File: tests/Blokemon.Game.Tests/TriggerTimingTests.fs
namespace Blokemon.Game.Tests

open Blokemon.Game
open FsUnit
open TUnit.Core

type TriggerTimingTests() =

    [<Test>]
    member _.``a rule-box knockout should take both remaining bar chits and win the match``() =
        let engine = MatchScenario.Engine()

        let state =
            MatchScenario.BattleState
                "BLK-026"
                "BLK-151"
                [ "VIM-BEER"; "VIM-BEER"; "VIM-SOBER" ]
                97UL

        let firstPrize =
            MatchScenario.PlainCard
                "prize-1"
                "VIM-LAIRY"
                MatchScenario.FirstPlayer
                CardZone.BarChit
                0

        let secondPrize =
            MatchScenario.PlainCard
                "prize-2"
                "VIM-SOBER"
                MatchScenario.FirstPlayer
                CardZone.BarChit
                1

        let state = MatchScenario.WithCards state [ firstPrize; secondPrize ]
        let state = MatchScenario.WithBarChits state MatchScenario.FirstPlayer 2

        let applied =
            MatchScenario.Applied(
                engine.Apply(state, MatchScenario.AttackCommand state "BLK-026-B01")
            )

        (applied.Card firstPrize.Id).Zone |> should equal CardZone.Mitt
        (applied.Card secondPrize.Id).Zone |> should equal CardZone.Mitt

        (applied.Player MatchScenario.FirstPlayer).BarChitsRemaining |> should equal 0

        applied.Winner |> should equal (ValueSome MatchScenario.FirstPlayer)

    [<Test>]
    member _.``a would-be-knocked-out trigger should resolve before the knockout``() =
        let engine = MatchScenario.Engine()

        let state =
            MatchScenario.BattleState
                "BLK-076"
                "BLK-068"
                [ "VIM-LAIRY"; "VIM-SOBER"; "VIM-SOBER" ]
                0UL

        let applied =
            MatchScenario.Applied(
                engine.Apply(state, MatchScenario.AttackCommand state "BLK-076-B02")
            )

        let defender = applied.Card(CardInstanceId "defender")
        defender.Zone |> should equal CardZone.Oche
        defender.Damage |> should equal 170

    [<Test>]
    member _.``a damaged-active trigger should place its counters on the attacker before knockouts``
        ()
        =
        let engine = MatchScenario.Engine()
        let state = MatchScenario.BattleState "BLK-076" "BLK-107" [ "VIM-LAIRY" ] 107UL

        let applied =
            MatchScenario.Applied(
                engine.Apply(state, MatchScenario.AttackCommand state "BLK-076-B01")
            )

        (applied.Card(CardInstanceId "attacker")).Damage |> should equal 30

    [<Test>]
    member _.``a taken-face-down trigger should wait for the bar chit's owner and the policy should complete it``
        ()
        =
        let engine = MatchScenario.Engine()

        let state =
            MatchScenario.BattleState
                "BLK-003"
                "BLK-001"
                [ "VIM-BLAZED"; "VIM-BLAZED"; "VIM-SOBER" ]
                0UL

        let triggeredPrize =
            MatchScenario.PlainCard
                "triggered-prize"
                "BLK-113"
                MatchScenario.FirstPlayer
                CardZone.BarChit
                0

        let extraPrize =
            MatchScenario.PlainCard
                "extra-prize"
                "VIM-LAIRY"
                MatchScenario.FirstPlayer
                CardZone.BarChit
                1

        let defenderBench =
            MatchScenario.PlainCard
                "defender-bench"
                "BLK-004"
                MatchScenario.SecondPlayer
                CardZone.Booth
                -1

        let state =
            MatchScenario.WithCards state [ triggeredPrize; extraPrize; defenderBench ]

        let state = MatchScenario.WithBarChits state MatchScenario.FirstPlayer 2

        let attacked =
            MatchScenario.Applied(
                engine.Apply(state, MatchScenario.AttackCommand state "BLK-003-B01")
            )

        attacked.Phase |> should equal MatchPhase.AwaitingTriggerChoice
        attacked.PendingBarChits.Count |> should equal 1

        let cpu = DeterministicCpu()

        let decision =
            MatchScenario.Chosen(cpu.Choose(engine, attacked, MatchScenario.FirstPlayer))

        let resolved = MatchScenario.Applied(engine.Apply(attacked, decision.Command))

        decision.Kind |> should equal LegalActionKind.ResolveBarChitTrigger
        (resolved.Card triggeredPrize.Id).Zone |> should equal CardZone.Booth
        (resolved.Card extraPrize.Id).Zone |> should equal CardZone.Mitt
        resolved.Winner |> should equal (ValueSome MatchScenario.FirstPlayer)

    [<Test>]
    member _.``a knockout vim move should wait for its own owner before the knockout completes``() =
        let engine = MatchScenario.Engine()

        let state =
            MatchScenario.BattleState
                "BLK-003"
                "BLK-001"
                [ "VIM-BLAZED"; "VIM-BLAZED"; "VIM-SOBER" ]
                103UL

        let triggerSource =
            MatchScenario.PlainCard
                "trigger-source"
                "BLK-026"
                MatchScenario.SecondPlayer
                CardZone.Booth
                -1

        let movableVim =
            MatchScenario.AttachedCard
                "movable-vim"
                "VIM-BEER"
                MatchScenario.SecondPlayer
                CardZone.Attached
                -1
                (CardInstanceId "defender")

        let prize =
            MatchScenario.PlainCard "prize" "VIM-LAIRY" MatchScenario.FirstPlayer CardZone.BarChit 0

        let defender =
            { state.Card(CardInstanceId "defender") with
                Attachments = FrozenList<CardInstanceId>.Create movableVim.Id }

        let state =
            MatchScenario.WithCards state [ defender; triggerSource; movableVim; prize ]

        let state = MatchScenario.WithBarChits state MatchScenario.FirstPlayer 1

        let attacked =
            MatchScenario.Applied(
                engine.Apply(state, MatchScenario.AttackCommand state "BLK-003-B01")
            )

        attacked.Phase |> should equal MatchPhase.AwaitingTriggerChoice
        attacked.PendingKnockout.IsSome |> should be True

        (attacked.Card movableVim.Id).AttachedTo
        |> should equal (ValueSome(CardInstanceId "defender"))

        (attacked.Card(CardInstanceId "defender")).Zone |> should equal CardZone.Oche

        let cpu = DeterministicCpu()

        let decision =
            MatchScenario.Chosen(cpu.Choose(engine, attacked, MatchScenario.SecondPlayer))

        let resolved = MatchScenario.Applied(engine.Apply(attacked, decision.Command))

        decision.Kind |> should equal LegalActionKind.ResolveKnockoutTrigger

        (resolved.Card movableVim.Id).AttachedTo
        |> should equal (ValueSome triggerSource.Id)

        (resolved.Card(CardInstanceId "defender")).Zone
        |> should equal CardZone.EmptiesTray

        (resolved.Card prize.Id).Zone |> should equal CardZone.Mitt
